### ISSUE TITLE
gem CLI: --silent, --quiet, --no-verbose, etc. should not print changelog

### ISSUE THREAD

**Issue by johnnyshields** [rubygems/rubygems#4355]
This report is related to **the command line `gem`**. It may be a regression, I'm not sure.

* When running `gem update --system 3.1.4 --quiet --no-document` (also `--silent`, `--no-verbose`) the changelog is **not** printed.
* When running `gem update --system 3.2.8 --quiet --no-document` (also `--silent`, `--no-verbose`) the changelog **is** printed.

I'd like to be able to  suppress the changelog for my Docker, etc. builds as it just takes up space in the log.

I've tested on a Windows machine but I believe the behavior is platform independent.

The Dude abides by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md). 🕶️🥛

---

For reference:

```
$>gem update --system 3.1.4 --quiet --no-document

Updating rubygems-update
Successfully installed rubygems-update-3.1.4
Installing RubyGems 3.1.4
  Successfully built RubyGem
  Name: bundler
  Version: 2.1.4
  File: bundler-2.1.4.gem
Bundler 2.1.4 installed
RubyGems 3.1.4 installed
Regenerating binstubs

$> gem update --system 3.2.8 --quiet --no-document

Updating rubygems-update
Successfully installed rubygems-update-3.2.8
Installing RubyGems 3.2.8
  Successfully built RubyGem
  Name: bundler
  Version: 2.2.8
  File: bundler-2.2.8.gem
Bundler 2.2.8 installed
RubyGems 3.2.8 installed
Regenerating binstubs
Regenerating plugins

# 3.2.8 / 2021-02-02

## Bug fixes:

* Fix `gem install` crashing on gemspec with nil required_ruby_version.
  Pull request #4334 by pbernays

# 3.2.7 / 2021-01-26

## Bug fixes:

* Generate plugin wrappers with relative requires. Pull request #4317 by
  deivid-rodriguez

<etc etc>
```

**Comment by deivid-rodriguez** [comment]
Thanks for reporting this @johnnyshields!

@duckinator would you be able to have a look at this?

**Comment by simi** [comment]
Looking at the options, there are two ones related:

https://github.com/rubygems/rubygems/blob/6be472bbc1c3c9f7ca559ab075514d32bd76536a/lib/rubygems/command.rb#L594-L601

Using `quite` mode seems to disable just progress meter per https://github.com/rubygems/rubygems/blob/6be472bbc1c3c9f7ca559ab075514d32bd76536a/lib/rubygems/user_interaction.rb#L374-L383

@johnnyshields can you try with `--silent`?

**Comment by johnnyshields** [comment]
@simi I tried silent as well, it also still showed the log lines.

**Comment by simi** [comment]
@johnnyshields ok, if @duckinator is out of time, I can fix this.

**Comment by deivid-rodriguez** [comment]
@simi I'd say go ahead. I think @duckinator is quite busy.

**Comment by duckinator** [comment]
Hey, as @deivid-rodriguez pointed out I'm pretty busy atm, but some context in case it helps: #2907 is a related issue about `--silent` not working, and #3789  is where it was fixed.

It is also worth noting that, because of the nature of fixing this specific functionality, the change will [only help when updating _from_ versions with the fix](https://github.com/rubygems/rubygems/issues/2907#issuecomment-756957235). That caused a bit of confusion before. :sweat_smile: 

@johnnyshields if you happen to know what version you tried upgrading _from_ when using `--silent`, it could help narrow down whether it's a regression or _just_ that I missed `--quiet` before.

@simi I know for a fact that I did not change `--quiet`, because I didn't know it existed (whoops). If it needs to be weaved through things in the same way as `--silent` was, the most relevant area is likely [`update_rubygem_arguments` from `lib/rubygems/commands/update_command.rb`](https://github.com/duckinator/rubygems/blob/2e6ef11/lib/rubygems/commands/update_command.rb#L299).

**Comment by simi** [comment]
Thanks for brief introduction @duckinator.

I did some initial work to support --silent in #4365.

The question still is how should "quiet" mode affect the output of rubygems install/update. I did a quick look and changelog is printed only when actual version updated has happened and `--quiet` doesn't affect that.

## changelog from 3.1.3 to 3.1.4 update

```
[retro@retro  rubygems ((v3.1.4) %)]❤ gem -v
3.1.4
[retro@retro  rubygems ((v3.1.4) %)]❤ ruby setup.rb --previous-version=3.1.3 --no-document --quiet
  Successfully built RubyGem
  Name: bundler
  Version: 2.1.4
  File: bundler-2.1.4.gem
Bundler 2.1.4 installed
RubyGems 3.1.4 installed
Regenerating binstubs

=== 3.1.4 / 2020-06-03

Minor enhancements:

* Deprecate rubyforge_project attribute only during build
  time. Pull request #3609 by Josef Šimánek.
* Update links. Pull request #3610 by Josef Šimánek.
* Run CI at 3.1 branch head as well. Pull request #3677 by Josef Šimánek.
* Remove failing ubuntu-rvm CI flow. Pull request #3611 by
  Josef Šimánek.


------------------------------------------------------------------------------

RubyGems installed the following executables:
        /home/retro/.rubies/ruby-2.7.2/bin/gem
        /home/retro/.rubies/ruby-2.7.2/bin/bundle
```

## no changelog from 3.1.4 to 3.1.4 (same) update

```
[retro@retro  rubygems ((v3.1.4) %)]❤ gem -v
3.1.4
[retro@retro  rubygems ((v3.1.4) %)]❤ ruby setup.rb --previous-version=3.1.4 --no-document --quiet
  Successfully built RubyGem
  Name: bundler
  Version: 2.1.4
  File: bundler-2.1.4.gem
Bundler 2.1.4 installed
RubyGems 3.1.4 installed
Regenerating binstubs



------------------------------------------------------------------------------

RubyGems installed the following executables:
        /home/retro/.rubies/ruby-2.7.2/bin/gem
        /home/retro/.rubies/ruby-2.7.2/bin/bundle
```
---

As you can see, passing `--quiet` doesn't do any difference. So I think this is not actual regression.

`--no-verbose` is used to suppress additional optional output, which is not this case (changelog is not currently printed only in verbose mode).

`--quiet` seems to be related only for progress reporting (even the naming is confusing a little), which is not changelog case as well IMHO.

`--silent` should suppress any standard output, but it doesn't work for setup command (which is called internally in `setup.rb`/`gem update --system call`. I'm trying to fix that in https://github.com/rubygems/rubygems/pull/4365.

@johnnyshields does that make sense to you and would `--silent` support work for you in this usecase? Alternatively (if you would like to skip changelog part only, but keep the rest), we will need to add additional flag.

**Comment by johnnyshields** [comment]
@simi `silent` works for me. I would suggest that the other two should also suppress the changelog as it can be quite long, but that's up to you.

**Comment by deivid-rodriguez** [comment]
I just tried `--silent` and it worked just fine for me. I think the problem, as mentioned from @duckinator, is that you need to upgrade from a version that has the fix, and 3.1.x doesn't have the fix since it was released with 3.2.0.

**Comment by deivid-rodriguez** [comment]
So, I'd say we can close this ticket and the related PR. If you want `--quiet` to be more quiet, we can open a separate issue about it.

**Comment by deivid-rodriguez** [comment]
Got no feedback here, so closing.

**Comment by pboling** [comment]
Note: `--silent` does not work with `gem update --silent --system`.  It is extremely verbose.

**Comment by duckinator** [comment]
@pboling unfortunately that's the case, yeah. Due to how RubyGems upgrades itself, you need to be upgrading _from_ a version that fixes this. My understanding is the fix was included in RubyGems 3.2.0, so anything upgrading _from_ 3.2.0 or newer should work.

**Comment by pboling** [comment]
Finding this again years later is fun!  The solution to upgrade from an older, noisier, rubygems / bundler is to use `> /dev/null` like this:
```
gem update --system > /dev/null
```

**Comment by duckinator** [comment]
FWIW: [Ruby 3.0 shipped with RubyGems 3.2.3](https://www.ruby-lang.org/en/news/2020/12/25/ruby-3-0-0-released/), which means all [currently-maintained versions of Ruby](https://www.ruby-lang.org/en/downloads/branches/) ship with a version of RubyGems without this problem.

So at this point, if you encounter this issue, please be aware you're likely using an unmaintained version of Ruby. ❤️

**Comment by pboling** [comment]
I wish more engineering leaders cared that software is EOL.  The sad reality is that many of them are working in the wrong job.
